File: app/src/androidTest/java/com/readtracker/android/db/MigrationTests.java
package com.readtracker.android.db;

import android.test.AndroidTestCase;

import com.readtracker.android.TestUtils;

import java.util.List;

public class MigrationTests extends AndroidTestCase {
  private final static String DATABASE_NAME = "readtracker-test.db";

  private DatabaseManager mDatabaseManager;

  @Override protected void setUp() throws Exception {
    super.setUp();
    DatabaseHelper databaseHelper = new DatabaseHelper(getContext(), DATABASE_NAME,
        null, 11);
    getContext().deleteDatabase(DATABASE_NAME);
    mDatabaseManager = new DatabaseManager(databaseHelper);
  }

  public void test_migration_12() {
    Book book = TestUtils.buildBook();
    mDatabaseManager.save(book);

    Session brokenSession = new Session();
    brokenSession.setTimestampMs(11158586000L /* Sun, 10 May 1970 03:36:26 GMT, ms */);
    brokenSession.setBook(book);
    mDatabaseManager.save(brokenSession);

    Session probablyOkSession = new Session();
    probablyOkSession.setTimestampMs(894771386000L /* Sun, 10 May 1998 03:36:26 GMT, ms */);
    probablyOkSession.setBook(book);
    mDatabaseManager.save(probablyOkSession);

    DatabaseHelper.migrateVersion31Sessions(mDatabaseManager);

    List<Session> session = mDatabaseManager.getAll(Session.class);
    assertEquals(2, session.size());
    final long THRESHOLD_DATE = 31536000000L /* Fri, 01 Jan 1971 00:00:00 GMT */;
    assertTrue(session.get(0).getTimestampMs() >= THRESHOLD_DATE);
    assertTrue(session.get(1).getTimestampMs() >= THRESHOLD_DATE);
  }
}

File: app/src/androidTest/java/com/readtracker/android/db/MigrationTests.kt
package com.readtracker.android.db

import junit.framework.Assert.assertTrue
import src.buildRandomBook
import org.junit.Assert
import org.junit.Test


class MigrationTest : DatabaseTestBase() {

    /**
     * Migrate broken session timestamps from version 3.1
     *
     * Create a session with a timestamp < 1971, that will be corrupted once
     * it's saved in the buggy 3.1 version of OrmLite. It is expected that it
     * will be saved in seconds instead of milliseconds. Then we create another
     * session that we expect to be perfectly fine. We fix the broken data if we find any
     * using the migrateVersion31Sessions function. Then we assert the results.
     */
    @Test
    fun migrationTest_MigrateFrom11to12_BugFix() {
        val dbManager = getManagerOfTestDatabaseAtVersion(11)

        val testBook = buildRandomBook()
        dbManager.save(testBook)

        fun saveSessionWithTimestamp(sessionTimestampMs: Long) {
            val session = Session().apply {
                timestampMs = sessionTimestampMs
                book = testBook
            }
            dbManager.save(session)
        }

        saveSessionWithTimestamp(11158586000L)
        saveSessionWithTimestamp(894771386000L)

        DatabaseHelper.migrateVersion31Sessions(dbManager)

        val session = dbManager.getAll(Session::class.java)
        Assert.assertEquals(2, session.size)

        val thresholdDate = 31536000000L /* Fri, 01 Jan 1971 00:00:00 GMT */
        Assert.assertTrue(session[0].timestampMs >= thresholdDate)
        Assert.assertTrue(session[1].timestampMs >= thresholdDate)
    }
}
